File: ahlQuran_Front/lib/system/new_models/attendance.dart
import 'package:flutter/material.dart';

class Attendance {
  final int? id;
  final int studentId;
  final String date; // Format: DD-MM-YYYY
  final String status; // "present", "late", "absent", "excused"
  final String? notes;
  final DateTime? createdAt;
  final DateTime? updatedAt;

  Attendance({
    this.id,
    required this.studentId,
    required this.date,
    required this.status,
    this.notes,
    this.createdAt,
    this.updatedAt,
  });

  factory Attendance.fromJson(Map<String, dynamic> json) {
    return Attendance(
      id: json['id'],
      studentId: json['student_id'],
      date: json['date'],
      status: json['status'],
      notes: json['notes'],
      createdAt: json['created_at'] != null
          ? DateTime.parse(json['created_at'])
          : null,
      updatedAt: json['updated_at'] != null
          ? DateTime.parse(json['updated_at'])
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      if (id != null) 'id': id,
      'student_id': studentId,
      'date': date,
      'status': status,
      'notes': notes,
      if (createdAt != null) 'created_at': createdAt!.toIso8601String(),
      if (updatedAt != null) 'updated_at': updatedAt!.toIso8601String(),
    };
  }

  // Get Arabic status text
  String getStatusArabic() {
    switch (status) {
      case 'present':
        return 'حاضر';
      case 'late':
        return 'متأخر';
      case 'absent':
        return 'غائب';
      case 'excused':
        return 'غائب بعذر';
      default:
        return status;
    }
  }

  // Get status color
  Color getStatusColor() {
    switch (status) {
      case 'present':
        return const Color(0xFF4CAF50); // Green
      case 'late':
        return const Color(0xFFFFA726); // Orange
      case 'absent':
        return const Color(0xFFEF5350); // Red
      case 'excused':
        return const Color(0xFF42A5F5); // Blue
      default:
        return Colors.grey;
    }
  }

  Attendance copyWith({
    int? id,
    int? studentId,
    String? date,
    String? status,
    String? notes,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) {
    return Attendance(
      id: id ?? this.id,
      studentId: studentId ?? this.studentId,
      date: date ?? this.date,
      status: status ?? this.status,
      notes: notes ?? this.notes,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }
}

File: ahlQuran_Back/app/schemas/attendance.py
from pydantic import BaseModel, Field
from datetime import datetime
from typing import Optional
from enum import Enum


class AttendanceStatus(str, Enum):
    PRESENT = "present"
    LATE = "late"
    ABSENT = "absent"
    EXCUSED = "excused"
    
    def __str__(self):
        return self.value


class AttendanceBase(BaseModel):
    date: str = Field(..., pattern=r'^\d{2}-\d{2}-\d{4}$', example="17-12-2024")  # DD-MM-YYYY
    status: AttendanceStatus = Field(default=AttendanceStatus.PRESENT, example="present")
    notes: Optional[str] = Field(None, example="Good attendance")


class AttendanceCreate(AttendanceBase):
    student_id: int = Field(..., gt=0, example=1)


class AttendanceUpdate(BaseModel):
    status: Optional[AttendanceStatus] = None
    notes: Optional[str] = None


class AttendanceResponse(AttendanceBase):
    id: int
    student_id: int
    created_at: datetime
    updated_at: Optional[datetime]

    class Config:
        from_attributes = True


class AttendanceList(BaseModel):
    attendances: list[AttendanceResponse]
    total: int
